### ISSUE TITLE
Custom metric not summing to correct value

### ISSUE THREAD

**Issue by page1** [rstudio/keras#164]
I've got a custom metric defined as:
```
y_t <- function(y_true, y_pred){
  return(K$sum(y_true))
}
```
Compiled like this
```
model %>% compile(
  loss = 'binary_crossentropy',
  optimizer = optimizer_adam(),
  metrics = c('acc',
              'f' = f_score_beta,
              'precision' = precision,
              'recall' = recall,
              'yt' = y_t,
              'yp' = y_p)
  )
```
Fit to all y's set to 1 ` as.array(rep(1L, nrow(x_train)))`
```
history <- model %>% fit(
  as.matrix(x_train), as.array(rep(1L, nrow(x_train))), 
  epochs = FLAGS$train_num_epochs, batch_size = FLAGS$train_batch_size, 
  validation_data = list(as.matrix(x_validation),
                         as.array(rep(1L, nrow(x_validation))))
)
```
I would expect the metric to simply report the size of the test set. But instead I get a floating point number that's closer to my batch size, but still far enough off I'm not sure I'd call it rounding. With batch size of 1000 I'm getting my metric to return 991.4259, but the training data is much larger than this.

Anyone else have this experience? Is this expected? I thought `y_true` should be identical to the passed in target values.

Thanks for any insight

**Comment by jjallaire** [comment]
I am pretty sure that metrics are computed per-batch so it being close to your batch size makes sense here. As for the apparent rounding I'm not sure what's happening there.

**Comment by civalin** [comment]
I got pretty the same problem here. Even I using

```python
def pred_sum(y_true, y_pred):
    return K.sum(K.round(y_pred))  # use K.round() make all values is integer...
```

The display value still not a integer. E.g.,

- `pred_sum: 0.146` when `batch_size=32`.
- `pred_sum: 0.0023` when `batch_size=1`, 

I don't know how to explain it, any suggestion? The strange results let all custom metrics data are doubtful. or I'm do something wrong?

## Environment

- Python 3.6
- Keras 2.1.6
- tensorflow 1.7.0
- numpy 1.14.2

## Code Sample

Here is some sample code.

```python3
from keras import Sequential
from keras.layers import Dense
import keras.backend as K


def true_sum(y_true, y_pred):
    return K.sum(y_true)

def pred_sum(y_true, y_pred):
    return K.sum(K.round(y_pred))

model = Sequential()
model.add(Dense(256, kernel_initializer='he_normal', activation='relu', input_shape=(6,)))
model.add(Dense(64, kernel_initializer='he_normal', activation='relu'))
model.add(Dense(1, activation='sigmoid'))
model.compile(optimizer='adam',
              loss='binary_crossentropy',
              metrics=[
                  'binary_accuracy',
                  true_sum,
                  pred_sum]
             )

scaled_inputs = ...  # not important
outputs = ...        # look like `np.array([[0], [1], ...])`

options = {'x': scaled_inputs,
           'y': outputs,
           'verbose': 2,
           'batch_size': 32,
           'epochs': 10,
           'validation_split': 0.1}
model.fit(**options)
```

The output is:

```
Using TensorFlow backend.

Train on 100505 samples, validate on 11168 samples
Epoch 1/10
 - 6s - loss: 0.5434 - binary_accuracy: 0.7591 - true_sum: 7.6895 - pred_sum: 0.2019 - val_loss: 0.5487 - val_binary_accuracy: 0.7517 - val_true_sum: 7.9484 - val_pred_sum: 0.0315
Epoch 2/10
 - 6s - loss: 0.5377 - binary_accuracy: 0.7598 - true_sum: 7.6894 - pred_sum: 0.1111 - val_loss: 0.5457 - val_binary_accuracy: 0.7516 - val_true_sum: 7.9484 - val_pred_sum: 0.0516
Epoch 3/10
 - 6s - loss: 0.5360 - binary_accuracy: 0.7599 - true_sum: 7.6895 - pred_sum: 0.1433 - val_loss: 0.5460 - val_binary_accuracy: 0.7522 - val_true_sum: 7.9484 - val_pred_sum: 0.3524
Epoch 4/10
 - 6s - loss: 0.5348 - binary_accuracy: 0.7600 - true_sum: 7.6892 - pred_sum: 0.2130 - val_loss: 0.5461 - val_binary_accuracy: 0.7521 - val_true_sum: 7.9484 - val_pred_sum: 0.1318
Epoch 5/10
 - 6s - loss: 0.5342 - binary_accuracy: 0.7598 - true_sum: 7.6892 - pred_sum: 0.2238 - val_loss: 0.5461 - val_binary_accuracy: 0.7530 - val_true_sum: 7.9484 - val_pred_sum: 0.6017
Epoch 6/10
 - 6s - loss: 0.5335 - binary_accuracy: 0.7599 - true_sum: 7.6894 - pred_sum: 0.2572 - val_loss: 0.5460 - val_binary_accuracy: 0.7530 - val_true_sum: 7.9484 - val_pred_sum: 0.6476
Epoch 7/10
 - 6s - loss: 0.5331 - binary_accuracy: 0.7602 - true_sum: 7.6895 - pred_sum: 0.2979 - val_loss: 0.5461 - val_binary_accuracy: 0.7520 - val_true_sum: 7.9484 - val_pred_sum: 0.1089
Epoch 8/10
 - 7s - loss: 0.5325 - binary_accuracy: 0.7604 - true_sum: 7.6895 - pred_sum: 0.3359 - val_loss: 0.5433 - val_binary_accuracy: 0.7532 - val_true_sum: 7.9484 - val_pred_sum: 0.4069
Epoch 9/10
 - 7s - loss: 0.5323 - binary_accuracy: 0.7606 - true_sum: 7.6893 - pred_sum: 0.3493 - val_loss: 0.5430 - val_binary_accuracy: 0.7537 - val_true_sum: 7.9484 - val_pred_sum: 0.2149
Epoch 10/10
 - 6s - loss: 0.5321 - binary_accuracy: 0.7603 - true_sum: 7.6895 - pred_sum: 0.3569 - val_loss: 0.5429 - val_binary_accuracy: 0.7522 - val_true_sum: 7.9484 - val_pred_sum: 0.3868
```
